server/src/main/java/org/apache/iotdb/db/mpp/execution/DataDriver.java

```diff
@@ -0,0 +1,324 @@
+import org.apache.iotdb.db.exception.query.QueryProcessException;
+import org.apache.iotdb.db.metadata.idtable.IDTable;
+import org.apache.iotdb.db.metadata.path.PartialPath;
+import org.apache.iotdb.db.mpp.buffer.ISinkHandle;
+import org.apache.iotdb.db.mpp.common.FragmentInstanceId;
+import org.apache.iotdb.db.mpp.operator.Operator;
+import org.apache.iotdb.db.mpp.operator.source.SourceOperator;
+import org.apache.iotdb.db.query.control.FileReaderManager;
+import org.apache.iotdb.tsfile.read.common.block.TsBlock;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.units.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static org.apache.iotdb.db.mpp.operator.Operator.NOT_BLOCKED;
+
+@NotThreadSafe
+public class DataDriver implements ExecFragmentInstance {
+
+  private static final Logger logger = LoggerFactory.getLogger(DataDriver.class);
+
+  private final Operator root;
+  private final ISinkHandle sinkHandle;
+  private final DataDriverContext driverContext;
+
+  private boolean init;
+  private boolean closed;
+
+  /** closed tsfile used in this fragment instance */
+  private Set<TsFileResource> closedFilePaths;
+  /** unClosed tsfile used in this fragment instance */
+  private Set<TsFileResource> unClosedFilePaths;
+
+  private final AtomicReference<SettableFuture<Void>> driverBlockedFuture = new AtomicReference<>();
+
+  public DataDriver(Operator root, ISinkHandle sinkHandle, DataDriverContext driverContext) {
+    this.root = root;
+    this.sinkHandle = sinkHandle;
+    this.driverContext = driverContext;
+    this.closedFilePaths = new HashSet<>();
+    this.unClosedFilePaths = new HashSet<>();
+    // initially the driverBlockedFuture is not blocked (it is completed)
+    SettableFuture<Void> future = SettableFuture.create();
+    future.set(null);
+    driverBlockedFuture.set(future);
+  }
+
+  @Override
+  public boolean isFinished() {
+    try {
+      boolean isFinished =
+          closed || (driverBlockedFuture.get().isDone() && root != null && root.isFinished());
+      if (isFinished) {
+        driverContext.finish();
+      }
+      return isFinished;
+    } catch (Throwable t) {
+      logger.error(
+          "Failed to query whether the data driver {} is finished", driverContext.getId(), t);
+      driverContext.failed(t);
+      close();
+      return true;
+    }
+  }
+
+  @Override
+  public ListenableFuture<Void> processFor(Duration duration) {
```

Comment:
In the `processFor` implementation, it should return a future with  `isCancelled() == true` if any error happens. So a single `NOT_BLOCKED` `ListenableFuture` may be insufficient.

Reply:
fixed